ENH enable multiview-oblique splitting & unlock multiclass restrictions

YuxinB opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines from scikit-learn: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are.

See https://github.com/neurodata/treeple/blob/main/CONTRIBUTING.md for more
information on contributing.

Thanks for contributing!
-->

mamamum committed:
make feature_combination into use

mamamum committed:
make feature_combination into use

YuxinB committed:
Merge pull request #353 from mamamum/main

Add BDD's changes to review

YuxinB committed:
Merge pull request #354 from neurodata/main

Update from Main

YuxinB committed:
Change feature_combination's default as None
in _multiview.py
Add non-axis-alighed weighted (fixing)

YuxinB committed:
Edit the splitter for multiview-oblique

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

adam2392 commented on treeple/tree/_oblique_splitter.pyx:
```diff
@@ -773,7 +774,7 @@ cdef class MultiViewSplitter(BestObliqueSplitter):
 
                     # here, axis-aligned splits are entirely weights of 1
-                    weight = 1  # if (rand_int(0, 2, random_state) == 1) else -1
+                    weight = 1 if (rand_int(0, 2, random_state) == 1) else -1
```
This will change the axis-aligned multi view-splitter I think(?)

YuxinB replied:
Oh, yes, I will change it back, it should be just 1, instead of 1 or -1

adam2392 reviewed (commented):
I think this looks correct, but I have not validated it, and also it's usually a bit hard to validate Cython code. 

I would recommend writing a small test Cython test class that you can invoke in Python, which then calls the `sample_proj_mat` for an input array. You can then visualize and verify that it will do what you want it to do for a set of features. You can then manually show that the MultiViewObliqueSplitter samples what you would expect to be the right projection matrix for multiple sets of features.

For example, see the `MultiViewSplitterTester` I wrote for test the MultiViewSplitter, and `plot_multiview_axis_aligned_splitter.py`.

YuxinB commented:
> I think this looks correct, but I have not validated it, and also it's usually a bit hard to validate Cython code.
> 
> I would recommend writing a small test Cython test class that you can invoke in Python, which then calls the `sample_proj_mat` for an input array. You can then visualize and verify that it will do what you want it to do for a set of features. You can then manually show that the MultiViewObliqueSplitter samples what you would expect to be the right projection matrix for multiple sets of features.
> 
> For example, see the `MultiViewSplitterTester` I wrote for test the MultiViewSplitter, and `plot_multiview_axis_aligned_splitter.py`.

Will do!

PSSF23 reviewed (commented):
I assume the changes in `forest.py` is to remove the multi-class restrictions? Then we can just remove the assertion line instead of commenting it out.

YuxinB committed:
Update Multiview-Oblique with uniform sampling
Add test for using same number of max_feature for each view

YuxinB committed:
Merge branch 'multiview-oblique' of https://github.com/neurodata/treeple into multiview-oblique

YuxinB commented:
@adam2392 Can you help me recheck with Multiview Oblique splitter when uniform sampling condition is true. I modified it to be more reasonable to achieve uniform sampling.

YuxinB committed:
Format

adam2392 commented:
> @adam2392 Can you help me recheck with Multiview Oblique splitter when uniform sampling condition is true. I modified it to be more reasonable to achieve uniform sampling.

Did a loose check, and it looks in the right direction. Will do an in-depth review after we can visually verify the projection matrix makes sense.

YuxinB committed:
Add MultiViewObliqueSplitterTester
Add example plot_multiview_oblique.py
Add different test for how to use max_feature

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

YuxinB committed:
Update Multiview Oblique Splitter with uniform sampling is true (Not general version)

YuxinB committed:
Merge branch 'multiview-oblique' of https://github.com/neurodata/treeple into multiview-oblique

YuxinB committed:
Update MultiviewObliqueSplitter when uniform sampling is true

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

adam2392 commented:
I'm unsure why the docs are broken, but it would be nice to be able to check the output visually here: https://output.circle-artifacts.com/output/job/96dcd4fa-4fbc-4b29-b358-314687b9af0b/artifacts/0/dev/use.html via the circleCI job: https://circleci.com/gh/neurodata/treeple/561. It makes reviewing easier/trivial even.

This is what I got though when I ran `plot_multiview_oblique.py` locally. To me this seems weird because your `feature_combinations` is 2, but on average, you're definitely sampling more than 2 feature indices per projection vector. Also, if this is an oblique combination, I'm unsure why the projection weights are only of value +1? For axis-aligned, we make them +1, but for oblique, we made them +/- 1, or 0. Without giving the code a deeper look, I conclude there's something not working. Open to being convinced tho.

![Figure_1](https://github.com/user-attachments/assets/859fa218-2f2b-431f-9e02-2e6b6feaee7d)

YuxinB commented:
When I directly output the projection matrix from `plot_multiview_oblique.py`, I got matrix:
<img width="247" alt="image" src="https://github.com/user-attachments/assets/ee45bd87-0fd8-4c8b-8e20-1edd0fa527d7" />
The feature_combination is around 2 on average. And there are -1s, 0s and 1s.

YuxinB committed:
Update Visualization test code

YuxinB commented:
Done with with visualization bug:
![image](https://github.com/user-attachments/assets/140be91b-8e86-4566-8a0d-94db9a000296)
![image](https://github.com/user-attachments/assets/75d70ac1-f6ec-42dc-90f8-208ffb7dad14)

YuxinB committed:
Merge branch 'multiview-oblique' of https://github.com/neurodata/treeple into multiview-oblique

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

YuxinB committed:
Update for Multiview Oblique Splitter

YuxinB committed:
Merge branch 'multiview-oblique' of https://github.com/neurodata/treeple into multiview-oblique

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

YuxinB committed:
Introduce `feature_combinations_per_set` to \
specify custom feature combinations for each set.

YuxinB committed:
Merge branch 'multiview-oblique' of https://github.com/neurodata/treeple into multiview-oblique
